support/cas-server-support-redis-ticket-registry/src/main/java/org/apereo/cas/ticket/registry/RedisTicketRegistry.java

```diff
@@ -172,7 +176,12 @@ public Ticket updateTicket(final Ticket ticket) {
     @Override
     public Ticket getTicket(final String ticketId, final Predicate<Ticket> predicate) {
         return FunctionUtils.doAndHandle(() -> {
-            val ticketPrefix = StringUtils.substring(ticketId, 0, ticketId.indexOf(UniqueTicketIdGenerator.SEPARATOR));
+            var decodedTicketId = ticketId;
+            val posSeparator = ticketId.indexOf(UniqueTicketIdGenerator.SEPARATOR);
+            if (posSeparator < 0 || posSeparator >= 10) {
```

Comment:
You're totally right, `10` is bad. In fact, this is not the code I had refactored. Please find the right one: https://github.com/apereo/cas/pull/5997/commits/2e1415b6bdbc32bdfb0f5c71fc22becbb3942ae3

I could replace `10` by `MAXIMUM_PREFIX_LENGTH` (I mean for `TGT-`, `AT-`, `CASMFA-`, ...) I found it better than explicitly checking all existing prefixes.

I think the problem is specific to Redis due to its specific logic: `CAS_TICKET:ticket_type:xxxx` as the key identifier, but this could maybe happen for some other ticket registries. I will check.

Reply:
OK. I was wrong.

I tested the property with Hazelcast and it fails:

```
ERROR [org.apereo.cas.ticket.DefaultTicketCatalog] - <Ticket definition for [eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCIsImtpZCI6ImE1NmMyYzI3LTE0ZGItNGU4OC1hN2IxLTI5Y2MyYTE0MTg3ZCJ9.ZXlKNmFYQWlPaUpFUlVZaUxDSmhiR2NpT2lKa2FYSWlMQ0psYm1NaU9pSkJNVEk0UTBKRExVaFRNalUySWl3aVkzUjVJam9pU2xkVUlpd2lkSGx3SWpvaVNsZFVJaXdpYTJsa0lqb2laall3TjJVeE1qZ3RaV1UwTWkwME56TTRMVGd5T0dJdE1qQXpNalZoWWpObFlqTXhJbjAuLm0wbGN3S2JFbXVJU2pCZXUtWjlya0EuQkJOM3gyREhiaU8tTTFLTlZXODVTeGxUWFhEZFBkQTRrN1BGYk5hNW5raEdUckFGMVEzV1hKTXdWX2NsWXp2MGFLOVlTUjFla3g5cFd6czNLTGZ5QUZ4NnlpMjRQT1g2MkZPaWV5OUwzWTl6TWNRTDUzbE55ZXpjUnZOXzJGNHguaERJZnJDM3A0ckxJNkh0dURQb3U1UQ.ojNRVpEMqk17wiEUpnoMwpHuqtw-fE84z01DdqimbOrA9Qslx6HZA4Akh8KHCAoKKQmL-n8mx5naXxZNK6TBag] cannot be found in the ticket catalog which only contains the following ticket types: [[TGT, ST, PT, PGT, TST]]>
```


I tested the property with Memcached as well and it fails afain:

```
ERROR [org.apereo.cas.ticket.registry.MemcachedTicketRegistry] - <Failed fetching [eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCIsImtpZCI6IjgzMmFlNDg2LTlkMjctNGVjYS1hMzdkLWU0YWUwNDhiNTAyMyJ9.ZXlKNmFYQWlPaUpFUlVZaUxDSmhiR2NpT2lKa2FYSWlMQ0psYm1NaU9pSkJNVEk0UTBKRExVaFRNalUySWl3aVkzUjVJam9pU2xkVUlpd2lkSGx3SWpvaVNsZFVJaXdpYTJsa0lqb2laalF3TW1ZM1pHVXRZbVV6WmkwME5EQTJMVGt3TW1FdE56TmxPR1V5TkRsaU5qY3dJbjAuLmVQcnJZTWNkcmxRYW0tbHRXVTVqZFEuMW9MWm1vb1R0N1d6eTlwSjhCcnBLY1B6SlFoaXBUMVE1Q3JCSy1ld2ZOajFxV0JIQTNudmZTeENreS10c1ZkZkhtSzduSmtPcGtrTUNManVzMXN0a2xpdkt3aHdTVmxNRDdYdXlseS1zTnBNVDlMejVnNVRtdnVJalBpMkhZNU4ubzBrSXFTenViVmQ0a3JSbXlxVy1UZw.nzXoqllGVEEupl2OjhzZqSkpcUCd0nzFMlA8bPdfhz9BEFdwArugHoBnqUXOmYk-6mMzWv0I-uyFR6PoJLvRag] >
ERROR [org.apereo.cas.ticket.registry.MemcachedTicketRegistry] - <Key is too long (maxlen = 250)
```

It seems the problem scope is bigger. I will work on a better solution.

Reply:
So I think this logic should be moved in the `public Ticket getTicket(final String ticketId)` method of the `AbstractTicketRegistry` component.

I don't think I convinced you with:

```java
var decodedTicketId = ticketId;
val posSeparator = ticketId.indexOf(UniqueTicketIdGenerator.SEPARATOR);
val separatorNotFound = posSeparator < 0;
val separatorFoundAfterPrefix = posSeparator >= 10;
val isIdentifierEncoded = separatorNotFound || separatorFoundAfterPrefix;
if (isIdentifierEncoded) {
    decodedTicketId = (String) protocolTicketCipherExecutor.decode(ticketId);
}
```

I guess you prefer something like:

```java
var decodedTicket = false;
for (var prefix : KNOWN_PREFIXES) {
    if (StringUtils.startsWith(ticketId, prefix)) {
        decodedTicket = true;
        break;
    }
}
val decodedTicketId = !decodedTicket ? (String) protocolTicketCipherExecutor.decode(ticketId) : ticketId;
```

Reply:
That is closer. Ideally, I would want to see:

```java
val decodedTicketId = protocolTicketCipherExecutor.decode(ticketId);
```

...and if the decoder can decode, it will. If it cannot or returns null, then we use the same ticket id.

You also need to look at where else this cipher is used because you run the risk of double encoding/decoding. If you look around, you likely find that tickets are decoded using this cipher in other areas. 

...finally, the easiest solution might would be to remove/deprecate the feature. I am not quite sure, at this point, what benefits it brings to the table if anything that and it's so rarely ever used that seems like a good candidate for removal.

Reply:
Yes, the best would be a smarter `protocolTicketCipherExecutor` preventing double encoding/decoding.

But I'm glad you raise the point of the usefulness of this setting. I don't see it. If I can steal an identifier to reuse it, why encoding it? By design, identifiers are opaque strings or signed/encrypted JWTs. I don't see the use of protecting them.

So I'm 100% in favor of deprecating and marking for removal this setting.

More than a matter of "easiest solution" (as I can always spend the required time needed to fix a problem), I think it's the most relevant solution.

I will submit a new PR to mark the setting for removal.

Reply:
=> https://github.com/apereo/cas/pull/6002

Reply:
I tend to agree, yes, but it's not about stealing an id per se. If you "cook up" your own token, you run the risk of hitting the registry with an invalid token and that would put load on the system only to come back and say "no ticket is invalid". The goal here was to prevent unnecessary calls to the registry if the ticket id is malformed or unknown such that we can be sure the ticket was created by CAS before we hit the registry, but even so, I am not sure I buy the advantages of that approach any more nowadays so it seems best that we deprecate and remove.